compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```
@@ -649,29 +649,334 @@ abstract class GenJSCode extends plugins.PluginComponent
+     *  1) The definition for the local variable that will hold the overload
+     *     resolution number.
+     *  2) The definitions of all local variables that are used as parameters
+     *     in all the constructors.
+     *  3) The overload resolution match/if statements. For each overload the
+     *     overload number is assigned and the parameters are cast and assigned
+     *     to their corresponding variables.
+     */
+    private def mkOverloadSelection(constructorTree: ConstructorTree,
+        overloadIdent: js.Ident, dispatchResolution: js.Tree,
+        isConstructor: Set[String])(implicit pos: Position): List[js.Tree]= {
+      if (!constructorTree.hasSubConstructors) {
+        dispatchResolution match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val refs = constructorTree.getParamRefs(method.name)
+            val rhss = stats.collect {
+              case js.VarDef(_, _, _, rhs) => rhs
+            }
+            zipMap(refs, rhss) { (ref, rhs) =>
+              js.VarDef(ref.ident, ref.tpe, mutable = false, rhs)
+            }
+
+          case js.ApplyStatic(_, mtd, js.This() :: Nil)
+              if ir.Definitions.isConstructorName(mtd.name) =>
+            Nil
+
+          case tree =>
+            List(tree)
+        }
+      } else {
+        val overloadRef = js.VarRef(overloadIdent)(jstpe.IntType)
+
+        def transformDispatch(tree: js.Tree): js.Tree = tree match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val num = constructorTree.getOverrideNum(method.name)
+            val refs = overloadRef :: constructorTree.getParamRefs(method.name)
+            val rhss = js.IntLiteral(num) :: stats.collect {
+              case js.VarDef(_, _, _, rhs) => rhs
+            }
+            val newStats = zipMap(refs, rhss)(js.Assign(_, _))
```

This looks fragile and obscure. The association between the order of `VarDef`s and the `getParamRefs` is unclear.

Consider
1. Separating `js.Assign(overloadRef, js.IntLiteral)` explicitly from the others
2. Using `stats.init` instead of `stats.collect` to get the list of all `js.VarDef`s. Especially if you assume there are exactly a list of `VarDef`s followed by one `js.ApplyStatic`. (currently other kinds of statements will be silently dropped)
3. Zipping `stats.init` with `getParamRefs`, and produce one `js.Assign` for each.